Add documentation to public API

michaelabenyohai opened the pull request:
Also rename methods/variables to be more consistent.

Commit:
Add documentation for MemoizationStore and implementations

Commit:
Add documentation for MemoizationStore and implementations

Commit:
Use single interface for Merkle leaf access rather than separate Function and Supplier

Commit:
Use single interface for Merkle leaf access rather than separate Function and Supplier

Commit:
Add documentation for MerkleTree classes

Commit:
Add documentation for MerkleTree classes

philandstuff commented:
needs rebase

Commit:
Resolve conflicts in consistency proof rename

Commit:
Merge branch 'documentation' of github.com:openregister/verifiable-log into documentation

Commit:
Rename tests

michaelabenyohai commented:
Rebased

philandstuff commented on src/test/java/uk/gov/verifiablelog/merkletree/MerkleTreeMockTests.java:
```diff
@@ -84,9 +84,9 @@ public void currentRoot_retrievesAndUsesMemoizationStoreHashes() {
+    public void auditProof_retrievesAndUsesMemoizationStoreHashes() {
         /**
-         *  Tree and memoization expectations for a consistency proof between trees of sizes 4 and 8
+         *  Tree and memoization expectations for a audit proof target node: 4, 8
```
should be `an audit`

philandstuff commented on src/main/java/uk/gov/verifiablelog/merkletree/MemoizationStore.java:
```diff
@@ -1,7 +1,24 @@
-    public void put(Integer start, Integer size, byte[] value);
-    public byte[] get(Integer start, Integer size);
+    /**
+     * Adds the root hash of a subtree to the set of known intermediate Merkle Tree root hashes.
```
I think it should be clear that `put` does not need to always add the root hash of a subtree, and that `get` can always return `null` if it feels like it

philandstuff commented on src/main/java/uk/gov/verifiablelog/merkletree/MerkleTree.java:
```diff
@@ -2,45 +2,73 @@
+ * An implementation of a Verifiable Log.
+ * @see <a href="https://tools.ietf.org/html/rfc6962">RFC 6962</a>
+ */
 public class MerkleTree {
```
we want to rename this at some point don't we?

Commit:
Renamed MerkleTree to VerifiableLog where relevant

Also restructured project into MerkleTree, VerifiableLog and DAO related packages and added test for MerkleUtil class.

michaelabenyohai commented:
@philandstuff I've addressed your comments in the latest commit.

philandstuff commented on src/main/java/uk/gov/verifiablelog/dao/memoization/MemoizationStore.java:
```diff
@@ -15,10 +16,11 @@
     byte[] get(Integer start, Integer size);
-}
-
+}
```
you should set intellij to not do this

philandstuff commented on src/main/java/uk/gov/verifiablelog/merkletree/MerkleUtil.java:
```diff
@@ -6,15 +6,17 @@
 /**
  * A utility class that provides commonly used operations when working with Merkle Trees.
```
I see this class as _not_ part of the public interface of the library, so we can maybe document that anything may change at any time

Commit:
Rename dao package to store and ensure Util class is package-private
